ccm_web/client/src/components/ConfirmationTable.tsx

```diff
@@ -0,0 +1,186 @@
+    }).sort((a, b) => (a.rowNumber < b.rowNumber ? -1 : 1))
+    )
+  }, [props.grades])
+
+  const renderTable = (): JSX.Element => {
+    const emptyRows = rowsPerPage - Math.min(rowsPerPage, tableRows.length - page * rowsPerPage)
+
+    const handleChangePage = (event: unknown, newPage: number): void => {
+      setPage(newPage)
+    }
+
+    const handleChangeRowsPerPage = (event: { target: { value: string } }): void => {
+      setRowsPerPage(parseInt(event.target.value, rowsPerPage))
+      setPage(0)
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    function TablePaginationActions (props: { count: any, page: any, rowsPerPage: any, onChangePage: any }): JSX.Element {
+      const classes = useStyles()
+      const theme = useTheme()
+      const { count, page, rowsPerPage, onChangePage } = props
+
+      const handleFirstPageButtonClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void => {
+        onChangePage(event, 0)
+      }
+
+      const handleBackButtonClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void => {
+        onChangePage(event, page - 1)
+      }
+
+      const handleNextButtonClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void => {
+        // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
+        onChangePage(event, props.page + 1)
+      }
+
+      const handleLastPageButtonClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void => {
+        onChangePage(event, Math.max(0, Math.ceil(count / rowsPerPage) - 1))
+      }
+
+      return (
+        <div className={classes.root}>
+          <IconButton
+            onClick={handleFirstPageButtonClick}
+            disabled={page === 0}
+            aria-label="first page"
+          >
+            {theme.direction === 'rtl' ? <LastPageIcon /> : <FirstPageIcon />}
+          </IconButton>
+          <IconButton onClick={handleBackButtonClick} disabled={page === 0} aria-label="previous page">
+            {theme.direction === 'rtl' ? <KeyboardArrowRight /> : <KeyboardArrowLeft />}
+          </IconButton>
+          <IconButton
+            onClick={handleNextButtonClick}
+            disabled={page >= Math.ceil(count / rowsPerPage) - 1}
+            aria-label="next page"
+          >
+            {theme.direction === 'rtl' ? <KeyboardArrowLeft /> : <KeyboardArrowRight />}
+          </IconButton>
+          <IconButton
+            onClick={handleLastPageButtonClick}
+            disabled={page >= Math.ceil(count / rowsPerPage) - 1}
+            aria-label="last page"
+          >
+            {theme.direction === 'rtl' ? <FirstPageIcon /> : <LastPageIcon />}
+          </IconButton>
+        </div>
+      )
+    }
+
+    TablePaginationActions.propTypes = {
+      count: PropTypes.number.isRequired,
+      onChangePage: PropTypes.func.isRequired,
+      page: PropTypes.number.isRequired,
+      rowsPerPage: PropTypes.number.isRequired
+    }
```

Comment:
Could you explain why `propTypes` is necessary, over just using TypeScript types with props in the component above? Something to do with this `ActionsComponent` thing?

Reply:
I based this on the typescript [example ](https://material-ui.com/components/tables/#CustomPaginationActionsTable.tsx) for Material-UI Table w/ pagination, which included it.  I'd guess there are improvements to be made with the table code.
As you mentioned elsewhere the confirmation and validation table compoments are quite similar.  That said, they are already based on a pretty base 'complex' component that is the Table, so I'm not entirely sure how much code sharing is really useful there, but it's worth looking into at some point.

The table code was more complicated than I expected so I did it as 2 separate components for now, and didn't stray from the example that much.

Reply:
Well fudge, I thought i was on the typescript example when I started.  I linked to the typescript example and saw interface TablePaginationActionsProps and confused it with the PropTypes version.. I'll take a look at this.

Reply:
Looks like you took care of this, thanks! Commented on the `package.json` about updating that.